File: python_dqi_cli/python_dqi_cli/schema.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any


@dataclass
class Timing:
    processing_ms: float
    model_inference_ms: float
    rss_delta_mb: float


@dataclass
class Scores:
    coverage_score: float
    structure_score: float
    noise_score: float
    info_density_score: float
    perplexity_score: None
    dqi_total: float


@dataclass
class DocumentResult:
    doc_id: str
    path: str
    features: dict[str, dict[str, Any]]
    scores: Scores
    timing: Timing

    def to_dict(self) -> dict[str, Any]:
        payload = asdict(self)
        payload["scores"] = asdict(self.scores)
        payload["timing"] = asdict(self.timing)
        return payload


@dataclass
class BenchmarkResult:
    mode: str
    docs_count: int
    rss_loaded_mb: float
    peak_rss_mb: float
    avg_rss_delta_per_doc_mb: float
    avg_processing_ms_per_doc: float
    p95_processing_ms_per_doc: float
    avg_model_inference_ms_per_doc: float
    total_batch_ms: float

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

File: src/schema.rs
use serde::Serialize;
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Serialize, Clone)]
pub struct Timing {
    pub processing_ms: f64,
    pub model_inference_ms: f64,
    pub rss_delta_mb: f64,
}

#[derive(Serialize, Clone)]
pub struct Scores {
    pub coverage_score: f64,
    pub structure_score: f64,
    pub noise_score: f64,
    pub info_density_score: f64,
    pub perplexity_score: Option<f64>,
    pub dqi_total: f64,
}

#[derive(Serialize, Clone)]
pub struct DocumentResult {
    pub doc_id: String,
    pub path: String,
    pub features: BTreeMap<String, Value>,
    pub scores: Scores,
    pub timing: Timing,
}

#[derive(Serialize)]
pub struct BenchmarkResult {
    pub mode: String,
    pub docs_count: usize,
    pub rss_loaded_mb: f64,
    pub peak_rss_mb: f64,
    pub avg_rss_delta_per_doc_mb: f64,
    pub avg_processing_ms_per_doc: f64,
    pub p95_processing_ms_per_doc: f64,
    pub avg_model_inference_ms_per_doc: f64,
    pub total_batch_ms: f64,
}
